server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -306,6 +352,73 @@ static boolean execute(SearchContext searchContext,
         }
     }
 
+     private static Query tryRewriteNumericLongOrDateSort(SearchContext searchContext, IndexReader reader) throws IOException {
+        if (searchContext.searchAfter() != null) return null;
+        if (searchContext.scrollContext() != null) return null;
+        if (searchContext.collapse() != null) return null;
+        if (searchContext.trackScores()) return null;
+        Sort sort = searchContext.sort().sort;
+        SortField sortField = sort.getSort()[0];
+        if (SortField.Type.LONG.equals(IndexSortConfig.getSortFieldType(sortField)) == false) return null;
```

Comment:
As discussed offline this should take into account `SortedNumericSortField`, see `IndexSortConfig#getSortFieldType` for an example.